socks.py

```diff
@@ -625,8 +626,17 @@ def _negotiate_HTTP(self, dest_addr, dest_port):
         # If we need to resolve locally, we do this now
         addr = dest_addr if rdns else socket.gethostbyname(dest_addr)
 
-        self.sendall(b"CONNECT " + addr.encode('idna') + b":" + str(dest_port).encode() +
-                     b" HTTP/1.1\r\n" + b"Host: " + dest_addr.encode('idna') + b"\r\n\r\n")
+        http_headers = [
+            b"Connect %s:%s HTTP/1.1" % (addr.encode('idna'), str(dest_port).encode()),
+            b"Host: %s" % dest_addr.encode('idna')
+        ]
+
+        if username and password:
+            http_headers.append(b"Proxy-Authorization: basic %s" % str(b64encode("%s:%s" % (username, password))))
```

Comment:
b64encode() does not accept text strings (in Python 3). You will have to explicitly encode the string.

In Python 2, %s is redundant with str(), and in Python 3.5, str() on a byte string is a bad idea, plus I don’t think %s will accept the result.